sdk/core/azure-core/src/base64.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/core/base64.hpp"
+#include "azure/core/platform.hpp"
+
+#if defined(AZ_PLATFORM_WINDOWS)
+#include <windows.h>
+#elif defined(AZ_PLATFORM_POSIX)
+#include <openssl/bio.h>
+#include <openssl/buffer.h>
+#include <openssl/evp.h>
+#endif
+
+#include <string>
+#include <vector>
+
+namespace Azure { namespace Core {
+
+#if defined(AZ_PLATFORM_WINDOWS)
+
+  std::string Base64Encode(const std::vector<uint8_t>& data)
+  {
+    std::string encoded;
+    // According to RFC 4648, the encoded length should be ceiling(n / 3) * 4
+    DWORD encodedLength = static_cast<DWORD>((data.size() + 2) / 3 * 4);
+    encoded.resize(encodedLength);
+
+    CryptBinaryToStringA(
```

Comment:
Check and throw when this return zero:
https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptbinarytostringa#return-value

Or how do we tell customers encoding failed?

Reply:
Great feedback. I intentionally left the implementation alone from what was in Storage, but we need more invalid tests and the error checking you mentioned.

I am also wondering if we should avoid relying on openssl as a dependency and OS-specific implementation for this, given Base64 encoding is relatively straightforward. I might revamp the whole implementation.

Reply:
I wish the docs stated **when** such API calls are expected to fail instead of just that they can fail. Generally encoding bytes to string shouldn't fail since almost all inputs are valid.

Reply:
The fails would usually come from unexpected inputs. Not the data, but the references or null pointers, etc.

Reply:
> Not the data, but the references or null pointers, etc.

In that case, since we wrote the code (and tested it), we should be calling it correctly :)
And if not, we need to fix the implementation. Since the failure won't be based on the user input passed into the function, we can assume that we are using the `CryptBinaryToStringA` API correctly.

That said, I will re-audit the implementation because I can see some edge cases that need to be addressed.